src/machine.rs

```diff
@@ -260,16 +260,38 @@ impl<'m> Machine<'m> {
         self.pid
     }
 
-    async fn send_action(&self, action: Action) -> Result<(), Error> {
-        let url: hyper::Uri = Uri::new(&self.config.host_socket_path, "/actions").into();
-        let json = serde_json::to_string(&action)?;
+    #[instrument(skip_all)]
+    async fn send_request(&self, url: hyper::Uri, body: String) -> Result<(), Error> {
+        let vm_id = self.config.vm_id();
+        trace!("{vm_id}: sending request to url={url}, body={body}");
+
         let request = Request::builder()
             .method(Method::PUT)
             .uri(url)
             .header("Accept", "application/json")
             .header("Content-Type", "application/json")
-            .body(Body::from(json))?;
-        self.client.request(request).await?;
+            .body(Body::from(body))?;
+
+        let resp = self.client.request(request).await?;
+
+        let status = resp.status();
+        if status.is_success() {
+            trace!("{vm_id}: request successful");
+        } else {
+            let body = hyper::body::to_bytes(resp.into_body()).await?;
+            let body = String::from_utf8_lossy(&body).into_owned();
+            trace!("{vm_id}: request failed: status={status}, body={body}");
+            let body = if body.is_empty() { None } else { Some(body) };
```

Comment:
you could avoid string conversion call if the bytes are empty and just print the prepared/created error?